### PR TITLE
Number equations across pages and by section (fix #3991, #4080)

### LINKED ISSUE #3991: Number equations within chapter
When the following code is included in `conf.py`, it allows figures and tables to be numbered within a chapter - the number becomes X.Y where X is the chapter number.

```
numfig = True

# number figures within section
numfig_secnum_depth = 1
```

It would be great if equations could also be numbered like this.

This comment https://github.com/sphinx-doc/sphinx/pull/3676#issuecomment-298872853 suggests that extending `numref` / `numfig` to cover equations is nontrivial. I've looked, but can't find enough information for me to have a go at doing it.

### PR DESCRIPTION
**Pull request by jahn**
### Feature or Bugfix
- Feature

### Purpose
Make equation numbers in html output unique across pages.

### Detail
This uses numfig for numbering and is only in effect if `numfig` and the new confval `math_numfig` are set.  Numbering is by section if `numfig_secnum_depth` is set.  The functionality is tested for each of the math backends.
This fix is admittedly rather hacky, but doing it properly seems to involve much deeper changes (see https://github.com/sphinx-doc/sphinx/pull/3676#issuecomment-298872853), and I feel a fair number of people would appreciate having this addressed (in particular preventing duplicate equation numbers on different pages), so I think it is still worth including.

### Relates
- #3991
- #4080 

Many thanks to @edoddridge for valuable input.

### CODE REVIEW

**Comment by hagenw** [comment]
Tested it, this indeed fixes #4080.

**Comment by mgeier** [comment]
Is it possible to get this reviewed and eventually merged?
I think it is a very useful feature!

File: doc/ext/math.rst [hunk 1]
```diff
@@ -44,6 +44,14 @@ or use Python raw strings (``r"raw"``).
 
    Example: ``'Eq.{number}'`` is rendered as ``Eq.10``
 
+.. confval:: math_numfig
+
+   If ``True``, displayed math equations are numbered across pages using numfig.
+   The :confval:`numfig` config value must be enabled and
+   :confval:`numfig_secnum_depth` is respected.  The ``:eq:`` role must be used
+   to refererence these equation numbers, not the ``:numref:`` role.
+   Default is ``False``.
```
Comment (tk0miya) [inline doc/ext/math.rst]: I think this is very useful, so I propose you to enable this by default. What do you think?

BTW, I think now this option is available only for HTML, not LaTeX output. Hence it should be mentioned.

**Review by tk0miya — CHANGES_REQUESTED** [review]
LGTM with nits.
Could you mention the limitation in docs please?

File: doc/ext/math.rst (hunk 1 above)
Comment (jfbu) [reply doc/ext/math.rst]: Perhaps make clearer option applies only to html like target (html, epub, singlehtml, ...). And ``math_numfig_for_eq`` might be clearer as name.

**Review by jfbu — CHANGES_REQUESTED** [review]
LGTM with nits, it appears to break singlehtml equation tags

**Comment by jfbu** [comment]
Here is screenshot for singlehtml test:
capture d ecran 2017-12-16 a 10 08 44

the equation tags are blank

**Comment by jfbu** [comment]
Regarding `numfig_secnum_depth` it is possible to implement LaTeX to obey it. Perhaps needs option ``latex_eq_obey_numfig_secnum_depth`` to activate or not the new feature. Currently, for `manual`, equations are numbered per "section" (which are `\chapter` in latex), and for `howto`, equations are numbered continuously across document.

File: doc/ext/math.rst (hunk 1 above)
Comment (jahn) [reply doc/ext/math.rst]: I have added the limitation to html-like output.
I'd be happy to enable it by default.
Happy also with the name change if others agree.

**Comment by jahn** [comment]
I think I have fixed the singlehtml output.  Let me know what the consensus on name and default value is and I'll implement it.

**Comment by jahn** [comment]
Regarding `numfig_secnum_depth`, I think latex currently doesn't respect it for figures and tables either.  My goal was for equation numbers to behave like figures and tables.  Consistency between html and latex can be achieved by setting `numfig_secnum_depth` to what latex uses.  I agree that it would be great to make this more flexible, though!

File: doc/ext/math.rst (hunk 1 above)
Comment (jfbu) [reply doc/ext/math.rst]: Can you use ``:rst:role:`eq` `` and ``:rst:role:`numref` `` mark-up?

About name change I have no strong opinion.

**Review by jfbu — APPROVED** [review]
Nice, thanks. I have briefly tested with html, singlehtml and epub and it appears to work fine.

As per LaTeX, indeed it does not obey numfig_secnum_depth already for figures and tables, so this is topic for separate PR.

**Comment by jfbu** [comment]
To avoid merge conflict because of CHANGES, the simplest is you do not worry about CHANGES and leave it as it is, and we update it after merging your work.

**Review by jfbu — COMMENTED** [review]
can you change the line at 

https://github.com/sphinx-doc/sphinx/pull/4166/files#diff-0f0370e5f85a0562dac5d13ac6905edfR114

because it appears obsolete at least for html like targets and latex, or I misunderstand what is meant by document.

@tk0miya any idea?

**Comment by jahn** [comment]
I am tempted to remove this sentence.  It now works across pages in HTML and Latex.  Texinfo, man pages and plain text do not show equation numbers at all.  The only place where I could find non-working equation numbers is in xml output.  I think fixing the xml output requires more knowledge of the inner workings of sphinx than I have.  Not sure it is worth it?
Sorry about the CHANGES mess, I realized my mistake too late.

File: doc/ext/math.rst (hunk 1 above)
Comment (tk0miya) [reply doc/ext/math.rst]: @jfbu this option also effect to block level equations (a.k.a displaymath node). So I feel `for_eq` is not good. Also, `math_` prefix describes clarify its purpose to me.

**Comment by tk0miya** [comment]
>because it appears obsolete at least for html like targets and latex, or I misunderstand what is meant by document.
>
>@tk0miya any idea?

I just posted #4313 now.

File: doc/ext/math.rst (hunk 1 above)
Comment (jfbu) [reply doc/ext/math.rst]: LGTM for `math_numfig` option name!

Comment (jahn) [reply doc/ext/math.rst]: Took care of :rst:role: markup.  Also changed default to True as requested.  I felt a slight rewording was appropriate in this case, which I applied.

**Review by jfbu — APPROVED** [review]
LGTM!

@tk0miya in reviewing this I have noticed that the ``:label:`` is not really good for display math. It tags only the first equation and there seems to be no way to tag other ones in the display. A ``:labels:`` seems missing. Besides, I look at latex output and complete change is needed to implement this. But this is other topic.

File: doc/ext/math.rst [hunk 2]
```diff
@@ -44,6 +44,14 @@ or use Python raw strings (``r"raw"``).
...
    Example: ``'Eq.{number}'`` is rendered as ``Eq.10``
 
+.. confval:: math_numfig
+
+   If ``True``, displayed math equations are numbered across pages in html and
+   related (epub, ...) output when :confval:`numfig` is enabled.
+   :confval:`numfig_secnum_depth` is respected.  The :rst:role:`eq` role must
+   be used to reference equation numbers, not the :rst:role:`numref` role.
+   Default is ``True``.
+
```
Comment (jfbu) [inline doc/ext/math.rst:54]: I forgot to say a ``.. versionadded:: 1.7`` should be added

**Comment by hagenw** [comment]
>@tk0miya in reviewing this I have noticed that the `:label:` is not really good for display math. It tags only the first equation and there seems to be no way to tag other ones in the display. A `:labels:` seems missing. Besides, I look at latex output and complete change is needed to implement this. But this is other topic.

I think the current behavior with `:label:` is not too bad and there is no urgent need to change this. You can easily decide yourself if you would like to have one label for multiple equations or multiple labels as well:

```rst
.. math::
    :label: single_label

    f(x) &= g(x) h(x) \\
    g(x) &= x \\
    h(x) &= 1 / x
```
eq_one_label

```rst
.. math::
    :label: label1

    f(x) = g(x) h(x)

.. math::
    :label: label2

    g(x) = x

.. math::
    :label: label3

    h(x) = 1 / x
```
eq_multiple_labels

I guess you would like to replace the last example by
```rst
..math::
    :labels: label1, label2, label3

    f(x) = g(x) h(x)
    g(x) = x 
    h(x) = 1 / x
```

File: doc/ext/math.rst (hunk 2 above)
Comment (jahn) [reply doc/ext/math.rst:54]: done

**Comment by jahn** [comment]
> I think the current behavior with `:label:` is not too bad and there is no urgent need to change this. You can easily decide yourself if you would like to have one label for multiple equations or multiple labels as well:

I think it would still be useful to have multiple labels.  This way one could align equations with separate labels which can sometimes improve readability.

**Comment by jahn** [comment]
Anything else I can do for this pull request?

**Comment by jfbu** [comment]
@hagenw  Yes I had in mind something like

```
.. math::
   :labels: foo1, foo2, foo3

   f(x) = g(x) h(x)

   g(x) = x 

   h(x) = 1 / x
```

but it had slipped my mind that `sphinx.ext.imgmath` generates a single image for the whole thing, hence in that case the idea is that mark-up would serve as shortcut to hand-written separated single equation.

For mathjax, with a single label Sphinx generates

```html
<div class="math" id="equation-foo">
<span class="eqno">(1)<a class="headerlink" href="#equation-foo" title="Permalink to this equation">¶</a></span>\[ \begin{align}\begin{aligned}f(x) = g(x) h(x)\\g(x) = x\\h(x) = 1 / x\end{aligned}\end{align} \]</div>
```

which is very similar to what latex builder does as well. Obviously there is some work here to get 3 labels, and again simplest is possibly some malaxing of rst doctree nodes  to actually split the whole into three separate things, but better surely would be to give benefits of aligned, if at all possible. I know LaTeX-amsmath, but not at all except as end user MathJaX which is a javascript rendering of some parts of TeX-amsmath syntax, hence no idea currently if we can get the label targets inside the display unit.

@jahn 

> Anything else I can do for this pull request?

Thanks for your work, LGTM! it motivated me for #4311 and once your work is merged I will when I get time do the same for LaTeX numbering of equations.

**Review by tk0miya — APPROVED** [review]
LGTM!

**Comment by tk0miya** [comment]
@jfbu I think the label is assigned to whole of equations-block. So it is not so bad. But I'm not a mathmatician, so good proposals are always welcome :-)

**Comment by tk0miya** [comment]
Now I merged this.
Thank you for good work!

**Comment by jahn** [comment]
Thanks for cleaning it up and including it!

**Comment by ar13pit** [comment]
@tk0miya the `align` and `align*` environments of AMSMath package allow to have multiple labels for a block of equations and they are really useful when representing a set of equations which also need to be aligned.

So implementing `:labels:` would be very helpful.

Regarding `sphinx.ext.imgmath` generating a single image, is it possible to break a block of equation into multiple images? 

Then `:labels:` would work for both `MathJax` and `sphinx.ext.imgmath`

**Comment by tk0miya** [comment]
@ar13pit why did you post here? Please file a new issue.

**Pull request merged**